File: src/tja/scrapers/canlii.nim
import std/[re, strutils, times]
import chronicles
from ../../tja/base_scraper import BaseScraper, initBaseScraper, makeRequest
from ../../tja/data_models import CaseData, initCaseData
from ../../tja/helpers import sanitizeText, normalizeCourtName


type
  CanLIIScraper* = ref object of BaseScraper

method baseUrl*(self: CanLIIScraper): string = "https://www.canlii.org"
method jurisdiction*(self: CanLIIScraper): string = "Canada"

proc newCanLIIScraper*(): CanLIIScraper = CanLIIScraper(initBaseScraper())

proc parseSearchHtml(self: CanLIIScraper; html: string; language = "en"): seq[CaseData] =
  var out: seq[CaseData] = @[]
  let resultRe = re"<div class=\"result\">([\s\S]*?)</div>"
  for rm in findAll(html, resultRe):
    let block = rm.group(0,1)
    let titleRe = re"<a[^>]*class=\"title\"[^>]*href=\"([^"]+)\"[^>]*>(.*?)</a>"
    let tm = titleRe.find(block)
    if tm.matchedLen == 0: continue
    var c = initCaseData()
    c.caseName = sanitizeText(block.substr(tm.group(0,2).a, tm.group(0,2).b))
    let href = block.substr(tm.group(0,1).a, tm.group(0,1).b)
    c.url = if href.startsWith("http"): href else: self.baseUrl() & href

    # meta: court and date
    let metaRe = re"<div class=\"resultmeta\">([\s\S]*?)</div>"
    let mm = metaRe.find(block)
    if mm.matchedLen > 0:
      let meta = sanitizeText(block.substr(mm.group(0,1).a, mm.group(0,1).b))
      let cd = re"([^,]+),\s*(\d{4}-\d{2}-\d{2})"
      let cdm = cd.find(meta)
      if cdm.matchedLen > 0:
        c.court = normalizeCourtName(meta.substr(cdm.group(0,1).a, cdm.group(0,1).b))
        try:
          c.date = parse("yyyy-MM-dd", meta.substr(cdm.group(0,2).a, cdm.group(0,2).b))
        except: discard
      # citations
      for cm in findAll(meta, re"\d{4}\s+[A-Z]+\s+\d+"):
        c.citations.add sanitizeText(meta.substr(cm.group(0,0).a, cm.group(0,0).b))

    # summary
    let sumRe = re"<div class=\"summary\">([\s\S]*?)</div>"
    let sm = sumRe.find(block)
    if sm.matchedLen > 0:
      c.summary = sanitizeText(block.substr(sm.group(0,1).a, sm.group(0,1).b))

    # caseId from URL
    let idm = re"/([^/]+)\.html$".find(c.url)
    if idm.matchedLen > 0:
      c.caseId = c.url.substr(idm.group(0,1).a, idm.group(0,1).b)

    c.jurisdiction = self.jurisdiction()
    out.add c
  return out

proc searchCases*(self: CanLIIScraper; query: string; startDate = ""; endDate = ""; court = ""; limit = 100; language = "en"): seq[CaseData] =
  var params: seq[(string,string)] = @[("resultCount", $(min(limit, 200))), ("sort","decisionDateDesc")]
  if query.len > 0: params.add ("text", query)
  if startDate.len > 0: params.add ("dateFrom", startDate)
  if endDate.len > 0: params.add ("dateTo", endDate)
  if court.len > 0: params.add ("tribunal", court)
  let url = self.baseUrl() & "/" & language & "/search/"
  let resp = self.makeRequest(url, params=params)
  result = self.parseSearchHtml(resp.body, language)
  if result.len > limit: result.setLen(limit)

proc getCaseById*(self: CanLIIScraper; caseId: string): CaseData =
  if caseId.len == 0: return initCaseData()
  var url: string
  if caseId.startsWith("http"):
    url = caseId
  elif '/' in caseId:
    url = self.baseUrl() & "/en/" & caseId & ".html"
  else:
    let cs = self.searchCases("citation:\"" & caseId & "\"", limit=1)
    if cs.len > 0: return cs[0]
    return initCaseData()
  let resp = self.makeRequest(url)
  var c = initCaseData()
  c.url = url
  let titleRe = re"<title>(.*?)</title>"; let mm = titleRe.find(resp.body)
  if mm.matchedLen > 0: c.caseName = sanitizeText(resp.body.substr(mm.group(0,1).a, mm.group(0,1).b))
  # court
  let courtRe = re"<span class=\"court\">([\s\S]*?)</span>"; let cm = courtRe.find(resp.body)
  if cm.matchedLen > 0:
    c.court = normalizeCourtName(sanitizeText(resp.body.substr(cm.group(0,1).a, cm.group(0,1).b)))
  # date
  let dateRe = re"<span class=\"date\">([\s\S]*?)</span>"; let dm = dateRe.find(resp.body)
  if dm.matchedLen > 0:
    let dt = sanitizeText(resp.body.substr(dm.group(0,1).a, dm.group(0,1).b))
    let dmatch = re"(\d{4}-\d{2}-\d{2})".find(dt)
    if dmatch.matchedLen > 0:
      try: c.date = parse("yyyy-MM-dd", dt.substr(dmatch.group(0,1).a, dmatch.group(0,1).b)) except: discard
  # citation
  let citRe = re"<span class=\"citation\">([\s\S]*?)</span>"; let cim = citRe.find(resp.body)
  if cim.matchedLen > 0:
    let ct = sanitizeText(resp.body.substr(cim.group(0,1).a, cim.group(0,1).b))
    if ct.len > 0: c.citations.add ct
  # full text heuristic
  let contRe = re"<div class=\"(documentcontent|document|content)\">([\s\S]*?)</div>"
  let contm = contRe.find(resp.body)
  if contm.matchedLen > 0:
    c.fullText = sanitizeText(resp.body.substr(contm.group(0,2).a, contm.group(0,2).b))
  # judges heuristic
  if c.fullText.len > 0:
    for jm in findAll(c.fullText, re"(?:Justice|Judge|J\.)\s+([A-Z][a-z]+(?:\s+[A-Z][a-z]+)*)"):
      let name = c.fullText.substr(jm.group(0,1).a, jm.group(0,1).b)
      if c.judges.len < 5 and not (name in c.judges): c.judges.add name
  # id
  let idm = re"/([^/]+)\.html$".find(url)
  if idm.matchedLen > 0: c.caseId = url.substr(idm.group(0,1).a, idm.group(0,1).b)
  c.jurisdiction = self.jurisdiction()
  return c

File: internal/scraper/jurisdictions/canlii.go
package jurisdictions

import (
	"context"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/PuerkitoBio/goquery"
	"github.com/gongahkia/kite/internal/scraper"
	"github.com/gongahkia/kite/pkg/errors"
	"github.com/gongahkia/kite/pkg/models"
)

// CanLIIScraper implements scraping for CanLII (Canadian Legal Information Institute)
type CanLIIScraper struct {
	*scraper.BaseScraper
	baseURL string
	client  *http.Client
}

// NewCanLIIScraper creates a new CanLII scraper
func NewCanLIIScraper() *CanLIIScraper {
	baseURL := "https://www.canlii.org"
	base := scraper.NewBaseScraper(
		"CanLII",
		"Canada",
		baseURL,
		15, // 15 requests per minute to be conservative
	)

	return &CanLIIScraper{
		BaseScraper: base,
		baseURL:     baseURL,
		client: &http.Client{
			Timeout: 30 * time.Second,
		},
	}
}

// SearchCases searches for cases matching the query
func (cs *CanLIIScraper) SearchCases(ctx context.Context, query scraper.SearchQuery) ([]*models.Case, error) {
	// Build search URL
	searchURL, err := cs.buildSearchURL(query)
	if err != nil {
		return nil, errors.ParsingError("failed to build search URL", err)
	}

	// Check robots.txt
	allowed, err := cs.BaseScraper.client.CheckRobots(ctx, "/en/search/")
	if err != nil || !allowed {
		return nil, errors.ErrRobotsDisallowed
	}

	// Wait for rate limit
	if err := cs.BaseScraper.client.rateLimiter.Wait(ctx); err != nil {
		return nil, errors.RateLimitError("rate limit exceeded")
	}

	// Make HTTP request
	req, err := http.NewRequestWithContext(ctx, "GET", searchURL, nil)
	if err != nil {
		return nil, errors.NetworkError("failed to create request", err)
	}

	req.Header.Set("User-Agent", cs.BaseScraper.client.userAgent)

	resp, err := cs.client.Do(req)
	if err != nil {
		return nil, errors.NetworkError("failed to fetch search results", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, errors.NetworkError(fmt.Sprintf("unexpected status code: %d", resp.StatusCode), nil)
	}

	// Parse HTML
	doc, err := goquery.NewDocumentFromReader(resp.Body)
	if err != nil {
		return nil, errors.ParsingError("failed to parse HTML", err)
	}

	// Extract cases from search results
	cases := make([]*models.Case, 0)

	doc.Find(".result").Each(func(i int, s *goquery.Selection) {
		if query.Limit > 0 && len(cases) >= query.Limit {
			return
		}

		caseData := cs.extractCaseFromSearchResult(s)
		if caseData != nil {
			cases = append(cases, caseData)
		}
	})

	return cases, nil
}

// GetCaseByID retrieves a specific case by its ID (e.g., "2023scc15")
func (cs *CanLIIScraper) GetCaseByID(ctx context.Context, caseID string) (*models.Case, error) {
	// CanLII case IDs are typically in format: 2023scc15
	// Parse to determine court and build URL
	caseURL := cs.buildCaseURL(caseID)

	// Check robots.txt
	allowed, err := cs.BaseScraper.client.CheckRobots(ctx, "/en/ca/")
	if err != nil || !allowed {
		return nil, errors.ErrRobotsDisallowed
	}

	// Wait for rate limit
	if err := cs.BaseScraper.client.rateLimiter.Wait(ctx); err != nil {
		return nil, errors.RateLimitError("rate limit exceeded")
	}

	// Make HTTP request
	req, err := http.NewRequestWithContext(ctx, "GET", caseURL, nil)
	if err != nil {
		return nil, errors.NetworkError("failed to create request", err)
	}

	req.Header.Set("User-Agent", cs.BaseScraper.client.userAgent)

	resp, err := cs.client.Do(req)
	if err != nil {
		return nil, errors.NetworkError("failed to fetch case", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusNotFound {
		return nil, errors.ErrNotFound
	}

	if resp.StatusCode != http.StatusOK {
		return nil, errors.NetworkError(fmt.Sprintf("unexpected status code: %d", resp.StatusCode), nil)
	}

	// Parse HTML
	doc, err := goquery.NewDocumentFromReader(resp.Body)
	if err != nil {
		return nil, errors.ParsingError("failed to parse HTML", err)
	}

	// Extract case details
	return cs.extractCaseDetails(doc, caseID, caseURL)
}

// GetCasesByDateRange retrieves cases within a date range
func (cs *CanLIIScraper) GetCasesByDateRange(ctx context.Context, startDate, endDate time.Time, limit int) ([]*models.Case, error) {
	query := scraper.SearchQuery{
		StartDate: &startDate,
		EndDate:   &endDate,
		Limit:     limit,
	}

	return cs.SearchCases(ctx, query)
}

// IsAvailable checks if CanLII is available
func (cs *CanLIIScraper) IsAvailable(ctx context.Context) bool {
	req, err := http.NewRequestWithContext(ctx, "HEAD", cs.baseURL, nil)
	if err != nil {
		return false
	}

	resp, err := cs.client.Do(req)
	if err != nil {
		return false
	}
	defer resp.Body.Close()

	return resp.StatusCode == http.StatusOK
}

// buildSearchURL builds the search URL with query parameters
func (cs *CanLIIScraper) buildSearchURL(query scraper.SearchQuery) (string, error) {
	params := url.Values{}

	if query.Query != "" {
		params.Set("search", query.Query)
	}

	if query.StartDate != nil && query.EndDate != nil {
		dateRange := fmt.Sprintf("%s:%s",
			query.StartDate.Format("2006-01-02"),
			query.EndDate.Format("2006-01-02"))
		params.Set("dateRange", dateRange)
	}

	searchURL := fmt.Sprintf("%s/en/search/?%s", cs.baseURL, params.Encode())
	return searchURL, nil
}

// buildCaseURL builds a case URL from a case citation (e.g., "2023scc15")
func (cs *CanLIIScraper) buildCaseURL(caseID string) string {
	// Simplified - in real implementation would parse the citation properly
	// Example: 2023scc15 -> /en/ca/scc/doc/2023/2023scc15/
	lowerID := strings.ToLower(caseID)

	// Extract year (first 4 digits)
	var year string
	for _, ch := range lowerID {
		if ch >= '0' && ch <= '9' {
			year += string(ch)
			if len(year) == 4 {
				break
			}
		}
	}

	// For now, assume SCC (Supreme Court of Canada) as example
	return fmt.Sprintf("%s/en/ca/scc/doc/%s/%s/", cs.baseURL, year, lowerID)
}

// extractCaseFromSearchResult extracts case data from a search result item
func (cs *CanLIIScraper) extractCaseFromSearchResult(s *goquery.Selection) *models.Case {
	c := models.NewCase()

	// Extract case name
	caseName := s.Find(".resultTitle a").Text()
	c.CaseName = strings.TrimSpace(caseName)

	// Extract case URL and ID
	caseURL, exists := s.Find(".resultTitle a").Attr("href")
	if exists {
		if !strings.HasPrefix(caseURL, "http") {
			caseURL = cs.baseURL + caseURL
		}
		c.URL = caseURL

		// Extract ID from URL
		parts := strings.Split(strings.Trim(caseURL, "/"), "/")
		if len(parts) > 0 {
			c.ID = parts[len(parts)-1]
		}
	}

	// Extract citation
	citation := s.Find(".resultCitation").Text()
	c.CaseNumber = strings.TrimSpace(citation)

	// Extract court and date from metadata
	metadata := s.Find(".resultMeta").Text()
	c.Summary = strings.TrimSpace(metadata)

	// Set basic metadata
	c.Jurisdiction = "Canada"
	c.SourceDatabase = "CanLII"
	c.ScrapedAt = time.Now()
	c.LastUpdated = time.Now()
	c.Language = "en"
	c.Status = models.CaseStatusActive

	return c
}

// extractCaseDetails extracts detailed case information from a case page
func (cs *CanLIIScraper) extractCaseDetails(doc *goquery.Document, caseID, caseURL string) (*models.Case, error) {
	c := models.NewCase()

	c.ID = caseID
	c.URL = caseURL

	// Extract case name
	caseName := doc.Find("h1.documentTitle").Text()
	c.CaseName = strings.TrimSpace(caseName)

	// Extract citation
	citation := doc.Find(".documentCitation").Text()
	c.CaseNumber = strings.TrimSpace(citation)

	// Extract court
	court := doc.Find(".court").Text()
	c.Court = strings.TrimSpace(court)

	// Extract date
	dateStr := doc.Find(".documentDate").Text()
	if dateStr != "" {
		// Try to parse various date formats
		formats := []string{"2006-01-02", "January 2, 2006", "02-01-2006"}
		for _, format := range formats {
			if date, err := time.Parse(format, strings.TrimSpace(dateStr)); err == nil {
				c.DecisionDate = &date
				break
			}
		}
	}

	// Extract docket number
	docket := doc.Find(".docketNumber").Text()
	c.Docket = strings.TrimSpace(docket)

	// Extract full text
	fullText := doc.Find(".documentContent").Text()
	c.FullText = strings.TrimSpace(fullText)

	// Extract judges
	doc.Find(".judge").Each(func(i int, s *goquery.Selection) {
		judge := strings.TrimSpace(s.Text())
		if judge != "" {
			c.Judges = append(c.Judges, judge)
		}
	})

	// Set metadata
	c.Jurisdiction = "Canada"
	c.SourceDatabase = "CanLII"
	c.ScrapedAt = time.Now()
	c.LastUpdated = time.Now()
	c.Language = "en"
	c.Status = models.CaseStatusActive

	return c, nil
}
